packages/gatsby/src/schema/__tests__/build-schema.js

```diff
@@ -1348,6 +1356,44 @@ describe(`Build schema`, () => {
         }"
       `)
     })
+
+    it(`handles merging types when implemented interface wasn't defined yet`, async () => {
+      createTypes(gql`
+        # create initial type composer
+        type TypeThatWillImplementInterface {
+          sharedField: String
+          originalField: String
+        }
+
+        # adjust type to implement not yet defined interface
+        # this will trigger type merging
+        type TypeThatWillImplementInterface implements CustomInterface {
+          sharedField: String
+          newField: String
+        }
+
+        # actually define interface (last)
+        interface CustomInterface {
+          sharedField: String
+        }
+      `)
+      const schema = await buildSchema()
+      expect(printType(schema.getType(`CustomInterface`)))
+        .toMatchInlineSnapshot(`
+        "interface CustomInterface {
+          sharedField: String
+        }"
+      `)
+
+      expect(printType(schema.getType(`TypeThatWillImplementInterface`)))
+        .toMatchInlineSnapshot(`
+        "type TypeThatWillImplementInterface implements CustomInterface {
+          sharedField: String
+          originalField: String
+          newField: String
+        }"
+      `)
+    })
```

Comment:
Main assertion is "implicit", just checking if `buildSchema` doesn't throw. Additional assertions making sure that type and interface is fully merged

Reply:
Added comment in code https://github.com/gatsbyjs/gatsby/pull/34089/commits/632ff6fdf263eb8c3c1f86f9bc9085510c075180